src/lib/rules/action.hh

```diff
@@ -24,7 +24,7 @@ public:
     virtual int check(const GameState& st) const = 0;
 
     // Apply action to game state.
-    virtual void apply(GameState& st) const = 0;
+    virtual void apply(GameState* st) const = 0;
```

Comment:
Why changing again? You changed the pointer into a ref in the last PR. And now it is back to pointer.

Personally I strongly prefer references because, pointers may imply ownership, and memory management. But some people have other reasons and prefer pointers.

I’m just puzzled to see it change again. Plus, it is not consistent with the `check` anymore (one uses a ref, the other a pointer).

Reply:
I reverted to using a pointer when I started refactoring code that uses this interface.

The rationale is that we should generally not have non-const reference arguments. This is confusing for the user of the interface, especially when you use check() and apply() close together, for example:

```
GameState st{};
ActionFoo action{};
action.check(st); // (1)
action.apply(st); // (2)
```

(1) All good, value-semantic for st, therefore st is a read-only argument, the reader can rightfully expect check() to take a const&

(2) Now in this line st is used with the same value syntax as (1), but the semantic is different: st in apply() is an output argument therefore it must have a reference semantic. The reader does not see the difference and does expect st to be modified. This is bad.

Now with a pointer the ambiguity is resolved by syntax:

```
GameState st{};
ActionFoo action{};
action.check(st); // value-semantic
action.apply(&st); // reference-semantic
```

As for ownership of the pointer in the method impl., we should use std::unique_ptr as the cannonical pointer ownership (and mem management, but that's the job of unique_ptr) marker. Raw pointers should only be used to convey the reference semantic.